internal/scheduler/unused/scheduler.go

```diff
@@ -0,0 +1,415 @@
+	item.priority = priority
+	heap.Fix(pq, item.index)
+}
+
+// Priority queue of jobs. Each queue has such a queue associated with it.
+type JobPQ []*Job
+
+// Job submitted to a queue.
+type Job struct {
+	Id string
+	// See events.proto.
+	Fragile bool
+	// Resource claims for this job, e.g., "CPU: 2000, RAM: "
+	// CPU is measured in millicores, RAM in bytes, and accelerators in millis.
+	Claims map[string]int64
+	// JobPQ priority.
+	priority float64
+	// JobPQ index. Maintained by the heap interface methods.
+	index int
+}
+
+func (pq JobPQ) Len() int {
+	return len(pq)
+}
+
+func (pq JobPQ) Less(i, j int) bool {
+	return pq[i].priority < pq[j].priority
+}
+
+func (pq JobPQ) Swap(i, j int) {
+	pq[i], pq[j] = pq[j], pq[i]
+	pq[i].index = i
+	pq[j].index = j
+}
+
+func (pq *JobPQ) Push(x interface{}) {
+	n := len(*pq)
+	item := x.(*Job)
+	item.index = n
+	*pq = append(*pq, item)
+}
+
+func (pq *JobPQ) Pop() interface{} {
+	old := *pq
+	n := len(old)
+	item := old[n-1]
+	old[n-1] = nil  // avoid memory leak
+	item.index = -1 // for safety
+	*pq = old[0 : n-1]
+	return item
+}
+
+// Update modifies the priority of an item in the queue.
+func (pq *JobPQ) Update(item *Job, priority float64) {
+	item.priority = priority
+	heap.Fix(pq, item.index)
+}
+
+// scheduleForever continually schedules jobs.
+func (s *Scheduler) scheduleForever() {
+	for {
+		err := s.scheduleOne()
+		if err != nil {
+			fmt.Printf("error: %s\n", err)
+		}
+		// TODO: Look at the error code to determine any changes that need to be made.
+	}
+}
+
+// scheduleOne schedules one job, or returns an error.
+func (s *Scheduler) scheduleOne() error {
+	queue, err := s.selectQueue()
+	if err != nil {
+		return err
+	}
+
+	job, err := queue.selectJob()
+	if err != nil {
+		return err
+	}
+
+	lease, err := s.suggestLease(queue, job)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("suggested lease: %s\n", lease)
+	return nil
+}
+
+// selectQueue returns the highest priority queue from the scheduler priority queue.
+// Does not remove the queue from the priority queue.
+func (s *Scheduler) selectQueue() (*Queue, error) {
+	if s.pq.Len() == 0 {
+		err := errors.Errorf("no queues available")
+		return nil, errors.WithStack(err)
+	}
+	return s.pq[0], nil
+}
+
+// selectJob returns the highest priority job from the queue priority queue.
+// Does not remove the queue from the priority queue.
+func (q *Queue) selectJob() (*Job, error) {
+	if q.Pq.Len() == 0 {
+		err := errors.Errorf("no jobs available")
+		return nil, errors.WithStack(err)
+	}
+	return q.Pq[0], nil
+}
+
+// suggestLease returns a suggested scheduling decision (i.e., a lease), or an error.
+// The caller is responsible for acting on that decision (or not) and updating the scheduler data structures accordingly.
+func (s *Scheduler) suggestLease(queue *Queue, job *Job) (*Lease, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if len(s.clusters) == 0 {
+		err := errors.Errorf("no clusters available")
```

Comment:
It feels a bit wrong to use errors to handle expected conditions (although maybe this is a go thing).